### PR TITLE
[GRPO] add gradient_checkpointing

### LINKED ISSUE #2671: "None of the inputs have requires_grad=True" with online DPO and GRPO
### Reproduction

Are online DPO and GRPO supposed to work with gradient checkpointing enabled?
I always get this warning when using them:
```/usr/local/lib/python3.11/dist-packages/torch/utils/checkpoint.py:87: UserWarning: None of the inputs have requires_grad=True. Gradients will be None```

And then the model doesn't seem to learn with a training loss that goes up and down, and the learning rate doesn't seem to have any impact.

Here is the notebook (simple code) to reproduce the error:
https://colab.research.google.com/drive/1Tb2m_EBdKuuELEEMkA7YYHmOIxozMBmu?usp=sharing

I tried many variations and first thought it was related to the use of an adapter but it isn't.

This notebook runs online DPO but I have the exact same problem with GRPO.
PS: use_vllm doesn't work with a peft config. In the same notebook, use_vllm=True and the peft_config trigger an error.

### System Info

Google Colab L4/A100

### Checklist

- [x] I have checked that my issue isn't already filed (see [open issues](https://github.com/huggingface/trl/issues?q=is%3Aissue))
- [x] I have included my system information
- [x] Any code provided is minimal, complete, and reproducible ([more on MREs](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any code provided is properly formatted in code blocks, (no screenshot, [more on code blocks](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any traceback provided is complete

### PR DESCRIPTION
**Pull request by kashif**
# What does this PR do?

fixes #2671

### CODE REVIEW

File: trl/trainer/grpo_trainer.py [hunk 1]
```diff
@@ -244,8 +244,24 @@ def __init__(
...
         if peft_config is not None:
+            if not is_peft_available():
+                raise ImportError("PEFT is required to use `peft_config`. Run `pip install peft`.")
             model = get_peft_model(model, peft_config)
 
+        # Enable gradient checkpointing if requested
+        if args.gradient_checkpointing:
+            # Ensure use_cache is disabled
+            if hasattr(model, "config"):
+                model.config.use_cache = False
```
Comment (qgallouedec) [inline trl/trainer/grpo_trainer.py:255]:
```suggestion
             model.config.use_cache = False
```

is should always be the case

File: trl/trainer/grpo_trainer.py [hunk 2]
```diff
@@ -244,8 +244,23 @@ def __init__(
...
+        # Enable gradient checkpointing if requested
+        if args.gradient_checkpointing:
+            # Ensure use_cache is disabled
+             model.config.use_cache = False
+
+            # Enable gradient checkpointing on the base model for PEFT
+            if peft_config is not None and hasattr(model.base_model, "gradient_checkpointing_enable"):
+                model.base_model.gradient_checkpointing_enable()
+            # Enable gradient checkpointing for non-PEFT models
+            elif hasattr(model, "gradient_checkpointing_enable"):
```
Comment (qgallouedec) [inline trl/trainer/grpo_trainer.py:260]:
```suggestion
            else:
```

always true

**Review by qgallouedec — APPROVED** [review]
